x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportEvaluateDataFrameAction.java

```diff
@@ -38,24 +39,64 @@ public TransportEvaluateDataFrameAction(TransportService transportService, Actio
                              ActionListener<EvaluateDataFrameAction.Response> listener) {
-        Evaluation evaluation = request.getEvaluation();
-        SearchRequest searchRequest = new SearchRequest(request.getIndices());
-        searchRequest.source(evaluation.buildSearch(request.getParsedQuery()));
-
-        ActionListener<List<EvaluationMetricResult>> resultsListener = ActionListener.wrap(
-            results -> listener.onResponse(new EvaluateDataFrameAction.Response(evaluation.getName(), results)),
+        ActionListener<List<Void>> resultsListener = ActionListener.wrap(
+            unused -> {
+                EvaluateDataFrameAction.Response response =
+                    new EvaluateDataFrameAction.Response(request.getEvaluation().getName(), request.getEvaluation().getResults());
+                listener.onResponse(response);
+            },
             listener::onFailure
         );
 
-        client.execute(SearchAction.INSTANCE, searchRequest, ActionListener.wrap(
-            searchResponse -> threadPool.generic().execute(() -> {
-                try {
-                    evaluation.evaluate(searchResponse, resultsListener);
-                } catch (Exception e) {
-                    listener.onFailure(e);
-                };
-            }),
-            listener::onFailure
-        ));
+        EvaluationExecutor evaluationExecutor = new EvaluationExecutor(threadPool, client, request);
+        // Add one task only. Other tasks will be added as needed during execution.
+        evaluationExecutor.add(evaluationExecutor.newTask());
+        evaluationExecutor.execute(resultsListener);
+    }
+
+    /**
+     * {@link EvaluationExecutor} class allows for serial execution of evaluation steps.
+     *
+     * Each step consists of the following phases:
+     *  1. build search request with aggs requested by individual metrics
+     *  2. execute search action with the request built in (1.)
+     *  3. make all individual metrics process the search response obtained in (2.)
+     *  4. check if all the metrics have their results computed
+     *      a) If so, call the final listener and finish
+     *      b) Otherwise, add another step to the queue
+     *
+     * To avoid infinite loop it is essential that every metric *does* compute its result at some point.
+     * */
+    private static final class EvaluationExecutor extends TypedChainTaskExecutor<Void> {
+
+        private final Client client;
+        private final EvaluateDataFrameAction.Request request;
+        private final Evaluation evaluation;
+
+        EvaluationExecutor(ThreadPool threadPool, Client client, EvaluateDataFrameAction.Request request) {
+            super(threadPool.generic(), unused -> true, unused -> true);
+            this.client = client;
+            this.request = request;
+            this.evaluation = request.getEvaluation();
+        }
+
+        private TypedChainTaskExecutor.ChainTask<Void> newTask() {
```

Comment:
Shall we call this `nextTask`? In a way this is like an iterator of sorts.

Reply:
Done.